hash/hash.go

```diff
@@ -0,0 +1,119 @@
+package hash
+
+import (
+	"bytes"
+	"crypto/subtle"
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/NebulousLabs/fastrand"
+	"golang.org/x/crypto/argon2"
+)
+
+var (
+	ErrInvalidHash               = errors.New("the encoded hash is not in the correct format")
+	ErrIncompatibleVersion       = errors.New("incompatible version of argon2")
+	ErrMismatchedHashAndPassword = errors.New("passwords do not match")
+
+	// This is the configuration of the argon2id hasher. For the moment we'll
+	// use the setup we had with Kratos. Later on we can move this to a config.
+
+	config = argon2Config{
+		SaltLength:  16,
+		Iterations:  2,
+		Memory:      131072,
+		Parallelism: 1,
+		KeyLength:   16,
+	}
+)
+
+type argon2Config struct {
+	SaltLength  uint32
+	Iterations  uint32
+	Memory      uint32
+	Parallelism uint8
+	KeyLength   uint32
+}
+
+// Generate hashes the given password with the default configuration.
+func Generate(password []byte) ([]byte, error) {
+	salt := fastrand.Bytes(int(config.SaltLength))
+	hash := argon2.IDKey(password, salt, config.Iterations, config.Memory, config.Parallelism, config.KeyLength)
+
+	var b bytes.Buffer
+	if _, err := fmt.Fprintf(
+		&b,
+		"$argon2id$v=%d$m=%d,t=%d,p=%d$%s$%s",
+		argon2.Version, config.Memory, config.Iterations, config.Parallelism,
+		base64.RawStdEncoding.EncodeToString(salt),
+		base64.RawStdEncoding.EncodeToString(hash),
+	); err != nil {
+		return nil, errors.AddContext(err, "failed to generate password hash")
+	}
+
+	return b.Bytes(), nil
+}
+
+// Compare verifies whether the given password matches the given hash.
+func Compare(password []byte, hash []byte) error {
+	// Extract the parameters, salt and derived key from the encoded password
+	// hash.
+	cf, salt, hash, err := decodeHash(string(hash))
+	if err != nil {
+		return err
+	}
+
+	// Derive the key from the other password using the same parameters.
+	otherHash := argon2.IDKey(password, salt, cf.Iterations, cf.Memory, cf.Parallelism, cf.KeyLength)
+
+	// Check that the contents of the hashed passwords are identical. Note
+	// that we are using the subtle.ConstantTimeCompare() function for this
+	// to help prevent timing attacks.
+	if subtle.ConstantTimeCompare(hash, otherHash) == 1 {
+		return nil
+	}
+	return ErrMismatchedHashAndPassword
+}
+
+// decodeHash is a helper method which extracts the configuration from the
+// encoded hash string and returns its parts.
+//
+// Example password record:
+// {"hashed_password":"$argon2id$v=19$m=131072,t=2,p=1$dwr95pEjaa7emZOu9bDAWw$eDQwOMoSyRmzyvpD/wwGBg"}
+func decodeHash(encodedHash string) (ac *argon2Config, salt, hash []byte, err error) {
```

Comment:
Follow Up: This method could use inline comments.

Reply:
I am not sure what I can add in those inline comments. The code is extremely straightforward and self-documenting.
It extracts the configuration for the `argon2` hashing (explained in the docstring) by extracting each bit of it in a separate block.

Example:  
```go
	salt, err = base64.RawStdEncoding.DecodeString(parts[4])
	if err != nil {
		return nil, nil, nil, err
	}
	ac.SaltLength = uint32(len(salt))
```
I can add a comment to this code which says "extracts the password salt" but that seems redundant to a level at which it reduces the redability of the code rather than enchancing it.

If you feel strongly about those comments I will add them but in my opinion they will not be an improvement.

Reply:
it is redundant if you leave out the why 😄 
Tell me the story, take me on this journey of decoding the hash.

imo inline comments should explain why we are doing each step as well as what each step is. I agree that each individual step is pretty straightforward in terms of what it is doing, but reviewing the code, or coming back to it to potentially edit it in 6 months, I can't pull the context of why each step is needed or if there are additionally order of operation concerns. 

Inline comments should save me the mental energy of having to read the code to understand the why.  

Example:
```
if version != argon2.Version {
		return nil, nil, nil, ErrIncompatibleVersion
}
```
I know this is checking for the correct version. But are there other considerations to consider with different versions? Do we need to offer compatibility code in the future if the version changes?

Reply:
Being able to understand the why through inline comments also helps reviewers provide better feedback or potential alternatives because it is clear what the developer's thought process was when writing the code.

Reply:
> Tell me the story, take me on this journey of decoding the hash.

Good point, you're absolutely right. Comments coming right up.